File: phase-c-information-systems-architecture/application-architecture/reward-bench/src/reward_bench/use_cases/bench.py
"""§7 `bench` — top-level composition.

    bench env cfg = argmaxBy (.score) (orchestrate env cfg)

Pure: any IO is the orchestrator's.
"""
from __future__ import annotations

from src.ports.orchestrator import Orchestrator
from src.reward_bench.entities.bench_config import BenchConfig
from src.reward_bench.entities.env import Env
from src.reward_bench.use_cases.best_submission import best_submission
from src.tier1.entities.submission import Submission


def bench(
    orchestrator: Orchestrator,
    env: Env,
    cfg: BenchConfig,
) -> Submission:
    return best_submission(orchestrator.orchestrate(env, cfg))

File: phase-c-information-systems-architecture/application-architecture/reward-bench/src/bench.erl
%% @doc §1 bench use case.
%%
%% bench :: env() -> bench_config() -> submission()
%% bench(Env, Cfg) = argmaxBy(.score)(orchestrator:orchestrate(Env, Cfg))
%%
%% Pure composition over the Orchestrator port.
-module(bench).
-export([bench/2]).

-include("records.hrl").

-spec bench(#env{}, #bench_config{}) -> #submission{}.
bench(Env, Cfg) ->
    case orchestrator:orchestrate(Env, Cfg) of
        [] ->
            #submission{body = <<>>, score = 0.0, walltime_sec = 0.0};
        [First | Rest] ->
            lists:foldl(fun pick_best/2, First, Rest)
    end.

pick_best(S, Acc) when S#submission.score > Acc#submission.score -> S;
pick_best(_S, Acc) -> Acc.
